portal-impl/src/com/liferay/portal/servlet/filters/virtualhost/VirtualHostFilter.java

```diff
@@ -312,18 +312,18 @@ protected void processFilter(
 					return;
 				}
 
+				String homeURL = PortalUtil.getRelativeHomeURL(
+					httpServletRequest);
+
 				if (Objects.equals(
 						group.getGroupKey(),
 						PropsValues.VIRTUAL_HOSTS_DEFAULT_SITE_NAME) &&
 					friendlyURL.equals(StringPool.SLASH) &&
-					!layoutSet.isPrivateLayout()) {
-
-					String homeURL = PortalUtil.getRelativeHomeURL(
-						httpServletRequest);
+					!layoutSet.isPrivateLayout() &&
+					Validator.isNotNull(homeURL) &&
+					!homeURL.equals(StringPool.SLASH)) {
```

Comment:
@holatuwol By changing the conditions of the if block here, you actually changed what's going to fall into the else block. I would like to be carefully about url generation as it is hard to test and tell what is correct or not. Could you please make this change by adding a new condition for the assign at:
if (Validator.isNotNull(homeURL) **&& !homeURL.equals(StringPool.SLASH)**) {
		friendlyURL = homeURL;
}

Reply:
@tinatian We actually need it to fall into the else block. If it doesn't, then the URL will stay "/", which doesn't address the issue.

Reply:
@holatuwol If this is the case, you commit message did not show what you want to do. Also, a change like this with implicit effect is risky, please try to make it clear by copying part of logic in else block to achieve your goal rather than making it fall into else block silently.

Please also add test to show what you want, this way we are clear about this change when we trace back later.

Tina.